topik/pemrograman-kompetitif-dasar/02-matematika-diskret-dasar/pemrograman-kompetitif-dasar_02-matematika-diskret-dasar.tex

```diff
@@ -11,87 +11,278 @@
+  function euclid(a,b : longint) : longint;
+  begin
+    if (b = 0) then begin
+      euclid := a;
+    end
+    else begin
+      euclid := euclid(b,a mod b);
+    end;
+  end;
+  \end{lstlisting}
+  \item Jika anda tertarik, anda dapat membaca lebih lanjut \textcolor{blue}{\href{https://en.wikipedia.org/wiki/Euclidean_algorithm}{di link wikipedia ini}.}   
 \end{itemize}
 \end{frame}
 
+\section{Generating Prime}
+\frame{\sectionpage}
+
 \begin{frame}
-\frametitle{Panduan Konten: Faktor Bilangan dan Sifatnya}
+\frametitle{Solusi Awal}
 \begin{itemize}
-  \item Jelaskan tentang konsep faktorisasi bilangan
-  \item Kaitkan dengan faktorisasi prima \& bilangan prima
-  \item Kaitkan dengan divisibility
+  \item Kita tahu bahwa bilangan prima merupakan bilangan yang hanya habis dibagi 1 dan dirinya sendiri.
+  \item Oleh karena itu, kita dapat melakukan \textit{looping} untuk mengecek apakah suatu bilangan prima atau tidak.
+  \item Maka, untuk tiap bilangan, kita coba-coba keterbagiannya dengan bilangan lain. Karena ada $N$ bilangan dan $N-2$ kemungkinan pembagi, maka kompleksitas solusi ini $O(N^{2})$.
 \end{itemize}
 \end{frame}
 
+\begin{frame}[fragile]
+\frametitle{Solusi Awal (lanj.)}
+Solusi:
+  \begin{lstlisting}
+  procedure naiveGeneration(n : longint);
+  var
+    i, j : longint;
+    prima : boolean;
+  begin
+    for i := 1 to n do begin
+        prima := true;
+        for j := 2 to n - 1 do begin
+          if (i mod j == 0) then begin
+            prima := false;
+          end;  
+        end;
+        if (prima) then begin
+          writeln(i, ' adalah bilangan prima');
+        end;
+    end;
+  end;  
+  \end{lstlisting}
+\end{frame}
+
+\begin{frame}
+\frametitle{Solusi $\sqrt{N}$}
+\begin{itemize}
+  \item Kita dapat mengobservasi bahwa jika $x = a * b$, dan $a \leq b$, maka $a \leq \sqrt{x}$ dan $b \geq \sqrt{x}$.
+  \item Oleh karena itu, kita cukup mengecek kemungkinan pembagi hingga $\sqrt{x}$.
+  \item Maka, kita dapat mengoptimasi solusi sebelumnya menjadi $O(N\sqrt{N})$.
+\end{itemize}
+\end{frame}
+
+\begin{frame}[fragile]
+\frametitle{Solusi $\sqrt{N}$ (lanj.)}
+Solusi:
+  \begin{lstlisting}
+  procedure optimizedGeneration(n : longint);
+  var
+    i, j : longint;
+    prima : boolean;
+  begin
+    for i := 1 to n do begin
+        prima := true;
+        for j := 2 to trunc(sqrt(i)) do begin
+          if (i mod j == 0) then begin
+            prima := false;
+          end;  
+        end;
+        if (prima) then begin
+          writeln(i, ' adalah bilangan prima');
+        end;
+    end;
+  end;  
+  \end{lstlisting}
+\end{frame}
+
+\begin{frame}
+\frametitle{Sieve of Erathostenes}
+\begin{itemize}
+  \item Terdapat solusi yang lebih optimal untuk membangkitkan bilangan prima, yaitu \newTerm{Sieve of Erathostenes}.
+  \item Ide utama dari algoritma ini adalah menandai tiap bilangan komposit. Hal ini dapat dilakukan dengan \textit{looping} dari tiap bilangan prima dan \textit{array} untuk menandai.
```

Comment:
wth... gua baca2 lagi... ini gak jelas banget algonya gimana :v... "loopting dari tiap bilangan prima dan array untuk menandai"? What the hell it supposed to mean? :v